### DISCUSSION TITLE
Bar rotation frequency as a function of time

### DISCUSSION THREAD

**Discussion by KaiOneWeaver** [jobovy/galpy#501] [Q&A] [answered]
Hi everyone.
Could you help me with the problem? How it is possible to make the potential of the bar rotate with a time-dependent frequency (Omegab(t)) in the Galpy. I reviewed all the wrappers, but I didn't find anything suitable, maybe I missed something.  However, I have seen a wrapper for time-dependent amplitude of potential (mass),.

Thank you in advance

**Comment by jobovy** [comment, marked as answer]
Hi!

There isn't currently any wrapper that allows the bar to rotate with a time-dependent frequency, but it wouldn't be too difficult to implement one starting from [`SolidBodyRotationWrapperPotential`](https://github.com/jobovy/galpy/blob/main/galpy/potential/SolidBodyRotationWrapperPotential.py). However, if you need to do many integrations, you'll want the speed-ups of the C code and then you would have to add a C version as well. Again, not difficult based on the constant OmegaB wrapper, but a bit of work.

However, you _can_ use a time-dependent frequency by using `galpy`'s new capability to work in rotating reference frames (see docs currently [here](https://docs.galpy.org/en/latest/orbit.html#new-in-v1-8-orbit-integration-in-non-inertial-frames)).  In [`NonInertialFrameForce`](https://docs.galpy.org/en/latest/reference/potentialnoninertialframe.html), you can specify that you want to frame to be rotating with a time-varying frequency (just set `Omega` and `Omegadot`; these can both be functions, so `Omegadot` doesn't have to be constant). Then you just define a bar potential that doesn't rotate such that you are integrating in the frame in which the bar is at rest. Pseudo-code for doing this if the bar has `Omegab(t) = t**2`:
```
from galpy.potential import MWPotential2014, NonInertialFrameForce, SoftenedNeedleBarPotential
barp= SoftenedNeedleBarPotential(omegab=0.) # non-rotating bar
nip= NonInertialFrameForce(Omega=lambda t: t**2.,Omegadot=lambda t: 2.*t)
o= Orbit(...) # Make sure to use initial conditions in the bar frame
ts= numpy.linspace(0.,100.,1001)
o.integrate(ts,MWPotential2014+barp+nip)
```
As I mention in the code, you have to make sure to initialize the `Orbit` instance in the bar frame, which typically means replacing `vT --> vT - Omegab(t0)*R`.

I know that the documentation on how to use `NonInertialFrameForce` is quite sparse at the moment, so please don't hesitate to follow up if you have additional questions on how this works. I'll be able to be most helpful if you send a brief, self-contained code example that I can try to get to run.

I'm also going to convert this to a discussion, because there isn't any issue with the code.

**Reply by KaiOneWeaver** [reply]
Thank you very much for the quick and comprehensive answer.
I'll try both options, but I like the first idea more, because then I don't have to worry about which report system (inertial or noninertial) the data is displayed. 
Have a nice day!

**Reply by KaiOneWeaver** [reply]
Dear Jo Bovy, 
Sorry, but i'm faced with misunderstanding.
I am working with Galpay version 1.7, even before adding the opportunity to work with a non-inertial frame.
I am trying to work with non axisymmetric 3d rotational potentials (bars)., Working in an inertial frame, I expected that when integrating orbits in such a potential, the Jacobi integral (E-Omega*L) would not be conserved. But it is conserved. Is it normal ?  Because it turns out that Lz is proportional to E  (Lz linearly connected with E) with respect to an inertial system. If this is correct, it is very unexpected. 
Thanks a lot!

**Reply by jobovy** [reply]
For a bar rotating with a fixed pattern speed, the Jacobi integral is conserved, but E and Lz are not separately conserved. The Jacobi integral is the energy in the rotating reference frame, which is conserved because the potential is static in that reference frame.

**Reply by KaiOneWeaver** [reply]
Thank you very much, I got it. 
For some reason, I incorrectly thought that in the ratio Ej=E-Omega*L=const, E - energy in the non-inertial system. But I was wrong, and now i understand, that E is energy in inertial system and then E=const +Omega*L,   which explains the linear relationship between Lz and E. 
I am very grateful to you.
